compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +446,77 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
```

Comment:
No, I don't think that's quite right. What we're concerned about is whether or not the operation can _block_. How it does so doesn't matter, the important thing is that all operations to be done directly in the AM handler must not be able to block.